### ISSUE TITLE
Upgrade from 2.2.2 to 2.2.5 caused loss of event track calls

### ISSUE THREAD

**Issue by platedbay** [segmentio/analytics-ruby#161]
Hi there,

We recently upgraded our rails application to use the most up to date version of the gem, `2.2.5`, from `2.2.2`. After some investigation on my end due to reports of events not being tracked properly, I discovered that reverting back to `2.2.2` has resolved the issues we were seeing. I wasn't able to determine specifically why this was occurring, just that controlling everything else and going back down to `2.2.2` meant our existing events were properly tracked.

Ruby version: `2.5.0`
Rails version: `5.2.0`

We noticed some pretty significant changes between these patch versions, so we were curious if there were any known issues or new ways things should be approached to properly identify users and track events? We haven't had to do much maintenance in that specific area, so we're a bit perplexed as to what could have happened.

Happy to provide any and more information to get to the bottom of this.

**Comment by f2prateek** [comment]
Hmm.. interesting - so far we haven't heard of this.  2.2.4.pre was in a pre-release state for 2 months and nobody had reported any issues. 

Most of the changes were around bug fixes/improving existing retry behaviour, so if anything you should have expected to see better deliverability.
We also added end to end tests as part of these changes - so I wonder if this is an issue that only exhibits itself under heavy load. 

Could you provide any other details - such as logs possibly? You can email us at friends@segment.com and cc prateek@segment.com

**Comment by coolnalu** [comment]
also seeing this.

**Comment by f2prateek** [comment]
Hey @rohitpaulk can you take a look using the version Ruby and rails @platedbay mentioned?

**Comment by f2prateek** [comment]
@platedbay @coolnalu can you make sure to set `on_error` handler as per https://segment.com/docs/sources/server/ruby/ and share some logs? We have an end to end test that verifies events are being received by Segment -  https://github.com/segmentio/analytics-ruby/blob/master/spec/segment/analytics/e2e_spec.rb; so I'm wondering if this something to how you've setup your apps.

**Comment by UlyssesInvictus** [comment]
@f2prateek Nothing is being logged, so it appears as the failure is silent, without triggering any errors. I'm looking into it further.

**Comment by UlyssesInvictus** [comment]
Also, I can confirm, wrt

> We also added end to end tests as part of these changes - so I wonder if this is an issue that only exhibits itself under heavy load.

That I can reproduce the error locally, on a tiny server, so it doesn't seem to be load based.

**Comment by f2prateek** [comment]
@UlyssesInvictus got it - if you are able to share a reproducible example, please do - or email us at friends@segment.com and cc prateek@segment.com if you want the code to be private!

**Comment by UlyssesInvictus** [comment]
Cool, @f2prateek ; unfortunately, I don't think we're actually doing anything very complicated in the Segment initializers themselves, so I suspect the actual issue is some conflict with other batching gems, so I'll have to get back to you later after I manage to spin up a new project and repro with a simpler gemset.

However, for now, here's the basic analytics code:

```
module Analytics
  Client = Segment::Analytics.new(
    write_key: SEGMENT_KEY,
    stub: Rails.env.test?,
    on_error: proc do |status, msg|
      Rails.logger.error "Segment call failed with status #{status}: #{msg}"
    end
  )
end
```

and then it's as simple as calling `Analytics::Client.track(...)` 🤷‍♂️ .

In fact, I've tried the following in console:

```
client = Segment::Analytics.new(write_key: SEGMENT_KEY)
client.track(user_id: "dummy user id", event: "SEGMENT TEST")
client.flush
```

and can confirm that the only way to get the event to log was to toggle between `2.2.5` and `2.2.2`.

Later tomorrow, I'll see what code from our internal initializers/configs I can give you to help debug.

**Comment by f2prateek** [comment]
Thanks @rohitpaulk! I also did a sanity check with the empty commit, as well as pushed a commit that doesn't call flush and only relies on the `max_queue_size` flag (https://github.com/segmentio/analytics-ruby/compare/prateek/test), which also passed the end to end test https://travis-ci.org/segmentio/analytics-ruby/builds/378872966. This also verifies that this doesn't come out of a bug due to some changes in the way we process events after it hits `api.segment.io`.

It does seem likely that this may be specific to the way some apps are setup (possibly a dependency conflict as @UlyssesInvictus suggested), as I am seeing customers successfully sending events with `2.2.5` of the library.

**Comment by platedbay** [comment]
Thanks for looking into all of this folks. I just emailed the emails above with our `Gemfile.lock`. Hopefully it helps.

Also, I do want to clarify that we were still seeing events being fired, just not ALL of them, which made it even more confusing.

**Comment by rohitpaulk** [comment]
>  I just emailed the emails above with our Gemfile.lock. Hopefully it helps.

Thanks :) I tried running this with the gems in the provided `Gemfile.lock`, but wasn't able to trigger the bug. 

> Also, I do want to clarify that we were still seeing events being fired, just not ALL of them, which made it even more confusing.

@platedbay: Are there any patterns to this? Do you see the same events being dropped every time?

**Comment by rohitpaulk** [comment]
The logging PR has now been merged. All logs from the library will be prefixed with `[analytics-ruby]`, so they should be easy to search for. 


 @platedbay, @UlyssesInvictus, @coolnalu: If possible, could you try running the latest master for a while and capture logs? We're particularly interested in logs like 'Retrying request', or 'message has exceeded size limit'.

```ruby
gem "analytics-ruby", :git => "https://github.com/segmentio/analytics-ruby", :branch => "master"
```

Also, please make sure that the Rails logger is configured to emit `debug` messages. [ref](http://guides.rubyonrails.org/debugging_rails_applications.html#log-levels)

```ruby
Rails.logger.level = 0
```

Here's a screenshot of how the logs look on a fresh rails app: 

![screen shot 2018-05-19 at 11 01 51 pm](https://user-images.githubusercontent.com/3893573/40271275-ae6e90fe-5bb8-11e8-9bac-fa5c4d977b91.png)

**Comment by UlyssesInvictus** [comment]
@rohitpaulk Apologies for the long delay on responding to this thread, I've been bogged down in work tickets.

I've upgraded our gemfile to build from master as requested

```
gem "analytics-ruby", :git => "https://github.com/segmentio/analytics-ruby.git", :branch => "master", require: 'segment'
```

and run a single analytics call in the rails console. Unfortunately, I can confirm that all that gets logged is

```
[...]
[analytics-ruby] Sending request for 1 items
=> true
```
and there's still no event being sent to our upstream analytics (NewRelic, in this particular instance).

I confirmed again that merely switching the gem version back to `2.2.2`, running another bundle install, (and no other code changes), leads to events being logged again.

**Comment by UlyssesInvictus** [comment]
Here's the full body of the analytics object:

```
=> #<Segment::Analytics:0x00007febea70b880
 @client=
  #<Segment::Analytics::Client:0x00007febea70b858
   @max_queue_size=10000,
   @options=
    {:write_key=>"...", :stub=>false, :on_error=>#<Proc:0x00007febea70b8d0@(pry):5>},
   @queue=#<Thread::Queue:0x00007febea70b768>,
   @worker=
    #<Segment::Analytics::Worker:0x00007febea70b6c8
     @batch=#<Segment::Analytics::MessageBatch:0x00007febea70b5d8 @json_size=0, @max_message_count=100, @messages=[]>,
     @lock=#<Thread::Mutex:0x00007febea70b588>,
     @on_error=#<Proc:0x00007febea70b8d0@(pry):5>,
     @queue=#<Thread::Queue:0x00007febea70b768>,
     @request=
      #<Segment::Analytics::Request:0x00007febea70b560
       @backoff_policy=
        #<Segment::Analytics::BackoffPolicy:0x00007febea70b510
         @attempts=0,
         @max_timeout_ms=10000,
         @min_timeout_ms=100,
         @multiplier=1.5,
         @randomization_factor=0.5>,
       @headers=
        {"Accept"=>"application/json", "Content-Type"=>"application/json", "User-Agent"=>"analytics-ruby/2.2.5"},
       @http=#<Net::HTTP api.segment.io:443 open=false>,
       @path="/v1/import",
       @retries=10>,
     @write_key="...">,
```

Additionally, calling flush after tracking doesn't help:
```
[4] pry(main)> client.flush
=> nil
```

**Comment by f2prateek** [comment]
hey @UlyssesInvictus, thanks for the info. If possible, are you able to inspect the raw HTTP request being made? If we can try to curl the request and reproduce the behaviour then we can isolate the problem a bit better. You can email the HTTP request to us at friends@segment.com and cc prateek@segment.com.

**Comment by UlyssesInvictus** [comment]
Not sure how I might do this, as it's not initially logged seems like something that takes place inside the Rails/Segment guts; any recommendations on what to monkeypatch so I can open up the request? (Outside of, IDK, hacking open our router, or something lol)

**Comment by f2prateek** [comment]
If this is locally on your computer, the easiest might be to use something like https://www.charlesproxy.com/, this will let you inspect the exact contents of the HTTP request happening.

If you'd like, feel free to shoot me an email at prateek@segment.com and we can setup a time to debug this issue over a call.

**Comment by rohitpaulk** [comment]
Thanks, @UlyssesInvictus. I expected this to have something to do with retries, but doesn't seem so. I'll add the HTTP status code to the logs too. 

Could you get us a copy of your `Gemfile` + `Gemfile.lock`? Send to prateek@segment.com or post here, either is fine :)

**Comment by UlyssesInvictus** [comment]
@f2prateek @rohitpaulk I've emailed Prateek the Gemfile as requested. Let me know when the HTTP code is added, and I'll try rerunning with that as well :)

**Comment by f2prateek** [comment]
thanks @UlyssesInvictus! we just merged the logging changes https://github.com/segmentio/analytics-ruby/pull/164, if you'd like to try that out as well.

**Comment by UlyssesInvictus** [comment]
`[analytics-ruby] Response status code: 200`

Looks like it succeeds? So I guess for some reason, downstream it fails to submit the logged event.

**Comment by f2prateek** [comment]
Interesting, I was really hoping to see a non 2xx response to quickly be able to isolate it to a client or server issue.

I think it'd really helpful to see an example of an HTTP request taht did succeed with a 200 response code, but still didn't show up in the debugger. Let me know if you'd like us to schedule a call.

**Comment by UlyssesInvictus** [comment]
Not sure exactly what

> I think it'd really helpful to see an example of an HTTP request taht did succeed with a 200 response code, but still didn't show up in the debugger. 

means. Could you clarify.

I'm once again busy with work, but I'll try to follow up within a reasonable amount of time.

**Comment by f2prateek** [comment]
Sorry - to clarify I meant it'd be really helpful to see the exact HTTP request being made that succeeded with a 200 response code (either using Charles or some other similar proxy tool), but events in that request did not show up in the Segment debugger or an integration connected to Segment.

**Comment by regentgal** [comment]
Just wanted to add that we are also seeing a loss of events moving from 2.2.2 to 2.2.5. Will try out some of the troubleshooting listed here and send results along in a bit.

**Comment by f2prateek** [comment]
@regentgal thanks - I think what would really help is you're able to reproduce this with something like [Charles proxy](https://www.charlesproxy.com/) running and send us the exact HTTP request being made by the library (my email is prateek@segment.com).

**Comment by regentgal** [comment]
@f2prateek I've setup the Charles proxy. However, looking at the gem code, I didn't see a way to pass proxy info into analytics-ruby that it would make it to through to request.rb and Net/Http, so Charles won't be able to pick it up. I can send along a capture from Wireshark if you are able to use that.

Just an FYI, running on the current master, I did not see any retry messages in the output. And the only change from when it works to when it doesn't is moving from 2.2.2 to 2.2.5.

```
# Gemfile.lock
GIT
  remote: https://github.com/segmentio/analytics-ruby.git
  revision: dcad49be324fc3043f9c79ff8d269e1d804b17ac
  branch: master
  specs:
    analytics-ruby (2.2.5)
      commander (~> 4.4)
```

```
[1] pry(main)> ENV['SEGMENT_WRITE_KEY'].length
=> 32
[2] pry(main)> seg_client = Segment::Analytics.new(
[2] pry(main)*   write_key: ENV['SEGMENT_WRITE_KEY'],
[2] pry(main)*   on_error: proc { |_status, msg| Rails.logger("Segment Analytics Error - #{msg}") }
[2] pry(main)* )
=> #<Segment::Analytics:0x007fabf001a1c8
 @client=
  #<Segment::Analytics::Client:0x007fabf001a1a0
   @max_queue_size=10000,
   @options={:write_key=>"mStyq4k8kb9SqH9uJMn85900oltwvqja", :on_error=>#<Proc:0x007fabf001a240@(pry):4>},
   @queue=#<Thread::Queue:0x007fabf001a0d8>,
   @worker=
    #<Segment::Analytics::Worker:0x007fabf0019f98
     @batch=#<Segment::Analytics::MessageBatch:0x007fabf0019e30 @json_size=0, @max_message_count=100, @messages=[]>,
     @lock=#<Thread::Mutex:0x007fabf0019de0>,
     @on_error=#<Proc:0x007fabf001a240@(pry):4>,
     @queue=#<Thread::Queue:0x007fabf001a0d8>,
     @request=
      #<Segment::Analytics::Request:0x007fabf0019db8
       @backoff_policy=
        #<Segment::Analytics::BackoffPolicy:0x007fabf0019d68
         @attempts=0,
         @max_timeout_ms=10000,
         @min_timeout_ms=100,
         @multiplier=1.5,
         @randomization_factor=0.5>,
       @headers=
        {"Accept"=>"application/json", "Content-Type"=>"application/json", "User-Agent"=>"analytics-ruby/2.2.5"},
       @http=#<Net::HTTP api.segment.io:443 open=false>,
       @path="/v1/import",
       @retries=10>,
     @write_key="mStyq4k8kb9SqH9uJMn85900oltwvqja">,
   @worker_mutex=#<Thread::Mutex:0x007fabf0019fc0>,
   @write_key="mStyq4k8kb9SqH9uJMn85900oltwvqja">>
[3] pry(main)> seg_client.track(user_id: User.first.id, event: "Testing")
[analytics-ruby] Sending request for 1 items
=> true
[4] pry(main)> [analytics-ruby] Response status code: 200
[4] pry(main)>
[5] pry(main)> seg_client.flush
=> nil
```

The only tagged log messages:
```
18:20:04 web.1         | [analytics-ruby] Sending request for 1 items
18:20:05 web.1         | [analytics-ruby] Response status code: 200
```

**Comment by f2prateek** [comment]
hey @regentgal; thanks this is helpful. Could you send me an email at prateek@segment.com? I'd love to setup a call to debug this together in person

**Comment by rohitpaulk** [comment]
With the latest Gemfile provided, I was able to dial this down to a minimal case that I can reproduce on my local setup. Looks like this has something to do with the `oj` gem. I'm not sure exactly what yet, but I'll have an update soon.

The minimal Gemfile required to trigger the bug: 

```ruby
gem "analytics-ruby", :git => "https://github.com/segmentio/analytics-ruby.git", :branch => "master"
gem 'oj', '~> 3.3.9'
gem 'oj_mimic_json' # Remove this, and the bug disappears
gem 'rails', '~> 5.0.2'
```

Thanks a lot for bearing with us, @platedbay @coolnalu @UlyssesInvictus @regentgal :) The hard part is over, this should be fixed soon.

**Comment by f2prateek** [comment]
Just wanted to add a few more details on this issue. We've been investigating, and found that the root cause is the `oj` module. For some reason, when this gem is included and activated, the JSON in the http requests looks invalid.

What it sends (note the extra `hash`  and `json` field in the batch array).

```
{
    "sentAt": "2018-06-06T13:32:47.725-07:00",
    "batch": [{
        "hash": {
            "event": "TEST",
            "userId": "rohitpaulk@gmail.com",
            "anonymousId": null,
            "context": {
                "library": {
                    "name": "analytics-ruby",
                    "version": "2.2.5"
                }
            },
            "options": null,
            "integrations": null,
            "properties": {
                "iteration_number": "iteration-1"
            },
            "messageId": "2db08051-3a91-4fc0-810d-633cf26f6d7b",
            "timestamp": "2018-06-06T13:32:45.984-07:00",
            "type": "track"
        },
        "json": "{\"event\":\"TEST\",\"userId\":\"rohitpaulk@gmail.com\",\"anonymousId\":null,\"context\":{\"library\":{\"name\":\"analytics-ruby\",\"version\":\"2.2.5\"}},\"options\":null,\"integrations\":null,\"properties\":{\"iteration_number\":\"iteration-1\"},\"messageId\":\"2db08051-3a91-4fc0-810d-633cf26f6d7b\",\"timestamp\":\"2018-06-06T13:32:45.984-07:00\",\"type\":\"track\"}"
    }]
}
```

What it should have sent instead:

```
{
    "sentAt": "2018-06-06T13:32:47.725-07:00",
    "batch": [{"event":"TEST","userId":"rohitpaulk@gmail.com","anonymousId":null,"context":{"library":{"name":"analytics-ruby","version":"2.2.5"}},"options":null,"integrations":null,"properties":{"iteration_number":"iteration-1"},"messageId":"2db08051-3a91-4fc0-810d-633cf26f6d7b","timestamp":"2018-06-06T13:32:45.984-07:00","type":"track"}]
}
```

We're still trying to figure out why, but it looks like the `oj` gem is doing something that interacts incorrectly with the JSON code in our library.

**Comment by f2prateek** [comment]
hey folks! we think we have a fix in https://github.com/segmentio/analytics-ruby/pull/166, could you try it out and let us know? We can get this merged and get the release train started for this fix.

**Comment by rohitpaulk** [comment]
`2.2.6.pre` has been released, which contains a fix for this issue.

**Comment by semanticart** [comment]
Hi everyone. We were hit with this bug as well. 😞 

Sadly we were affected _after_ the pre-release of 2.2.6 was already shipped. We just updated to the latest official release in ignorance of this issue.

Is 2.2.6.pre stable enough to ship as a proper release? Having this fixed in general release would prevent others from running into this problem. It seems that right now they have no way of knowing things are broken until they discover the gap in their reporting.

--

Does the new fix cause the `on_error` to be fired if the payload format isn't properly formatted? If not, is that on the roadmap? That might help prevent future silent failures of this sort.

**Comment by semanticart** [comment]
Another thought: should 2.2.5 be yanked?

**Comment by rohitpaulk** [comment]
Hey @semanticart, sorry you hit this 😓 

> Is 2.2.6.pre stable enough to ship as a proper release?

We don't know yet, but you could help out 🙂 Would really appreciate it if you could upgrade to `2.2.6.pre` and test - if it solves the issue, I'll promote it to `2.2.6`.

> Having this fixed in general release would prevent others from running into this problem. It seems that right now they have no way of knowing things are broken until they discover the gap in their reporting.

Totally agree 👍 We're just waiting on at least one confirmation before pushing the update. 

> Another thought: should 2.2.5 be yanked?

Yanking could cause problems with CI, builds etc for a lot of customers - I'd like to avoid this as far as possible. As far as I've seen, yanking is commonly done when you accidentally publish credentials, or push a gem with a wrong name - things like that. As per my understanding, it isn't common to yank releases because they contain bugs.

**Comment by rohitpaulk** [comment]
> Does the new fix cause the on_error to be fired if the payload format isn't properly formatted? If not, is that on the roadmap? That might help prevent future silent failures of this sort.

This is a good idea, but will require changes to the segment API too (it currently returns a 200 on a badly formatted payload). Deferring to @f2prateek here - looks like something we could consider in the future 🙂

**Comment by semanticart** [comment]
@rohitpaulk Reporting in on `2.2.6.pre`. It did not fix the issue for us. We didn't have time to try to debug why so we're rolled back to `2.2.2` for now. 

FWIW, our gemfile versions for oj and oj_mimic_json are as follows:
```
    oj (2.18.5)
    oj_mimic_json (1.0.1)
```

**Comment by f2prateek** [comment]
@semanticart  I don't think we'll yank 2.2.5 as it seems to be a specific issue only affecting when you have very specific dependencies installed. If it was something more global then we would reconsider. 

But I think we can do a better job of calling this out in the Changelog.

@rohitpaulk can you update the changelog and add a note along the lines of "There is a known issue using 2.2.5 with oj. Please test out 2.2.6.pre and hold off on using 2.2.5".

**Comment by rohitpaulk** [comment]
> @rohitpaulk Reporting in on 2.2.6.pre. It did not fix the issue for us. We didn't have time to try to debug why so we're rolled back to 2.2.2 for now.

Thanks for taking a look at this @semanticart 🙂 

The version of `oj` you're running (`2.18.5`) seems to be more than a year old - could you try updating to the latest and see if that fixes the issue? In our tests, we're testing against `3.6.2`.

Also, what Ruby version are you running? Any additional info that could help us track this down would be super helpful. There are instructions on how to enable debug logs above in https://github.com/segmentio/analytics-ruby/issues/161#issuecomment-390420445. 

Lastly, although this is unlikely - please double check that you're indeed running the updated version by inspecting `Segment::Analytics::VERSION`. 

> @rohitpaulk can you update the changelog and add a note along the lines of "There is a known issue using 2.2.5 with oj. Please test out 2.2.6.pre and hold off on using 2.2.5".

Will do 👍

**Comment by semanticart** [comment]
> The version of oj you're running (2.18.5) seems to be more than a year old - could you try updating to the latest and see if that fixes the issue? In our tests, we're testing against 3.6.2.

This seemed to fix the issue. We were locked to an older version of `oj` because we're currently on an older version of Rails but it looks like `oj` has been made more compatible since the last time we tried upgrading or some other confluence of events. Anyway, upgrading `oj` and using the pre-release looks good on staging.

I'll try it against prod tomorrow but I'd expect similar results.

There might be something to learn by trying the older `oj` version on your end that might inform a more graceful way to handle failures, but 🤷‍♂️ 

For the record, we're on ruby 2.5.1

Thanks

**Comment by semanticart** [comment]
I've shipped to prod and feel comfortable confirming that oj `3.6.3` and analytics-ruby `2.2.6.pre` work well together. We're sending events properly. 🎉

**Comment by rvirani1** [comment]
I just hit this bug without using `oj` as well. I use the `yajl-ruby` gem with the [compatibility feature](https://github.com/brianmario/yajl-ruby#json-gem-compatibility-api). None of my data was getting into Segment and there was really no way to know without finding this thread. I poked through the library source and Segment seemed to be returning 200s. Additionally, nothing negative was logged and the error handler never fired. This seems like a major risk to me. How would I ever know if some data was just getting dropped silently?

**Comment by jtoy** [comment]
seeing this issue as well

**Comment by sagan-boundless** [comment]
I know this issue is old/closed, but I am posting some info in case it helps others in the future.

We are using the gem `analytics-ruby` version 2.4.0 (https://github.com/segmentio/analytics-ruby)
We had the same problem recently, and our fix was to downgrade OJ.

The problem started when we upgraded OJ to version 3.14.3.
So we downgraded back to 3.10.14, and all was working fine again.

Hope this helps...
